### PR TITLE
State Read/Write/Clear range implementation

### PR DESCRIPTION
**Pull request by mitch-fuel**
Signed-off-by: Mitch Martin <mitch.martin@fuel.sh>

### CODE REVIEW

**Comment by Voxelot** [comment]
Moved to draft, as the PR isn't ready for review

File: fuel-core/src/database/vm_database.rs [hunk 1]
```diff
@@ -142,28 +158,1192 @@ impl InterpreterStorage for VmDatabase {
...
+            .unwrap();
+        let read_2 = read_db
+            .merkle_contract_state(&contract_id, &Bytes32::new(key_2))
+            .unwrap();
+
+        assert_eq!(pre_insert_read_0.is_none(), true);
+        assert_eq!(pre_insert_read_1.is_none(), true);
+        assert_eq!(pre_insert_read_2.is_none(), true);
+        assert_eq!(insert_status.is_none(), true);
+        assert_eq!(read_0.is_none(), false);
```
Comment (Dentosal) [inline fuel-core/src/database/vm_database.rs:756]: nit: Simplify to `assert!(pre_insert_read_0.is_none());`

Comment (mitch-fuel) [reply fuel-core/src/database/vm_database.rs]: fixed all

File: fuel-core/src/database/vm_database.rs [hunk 2]
```diff
@@ -142,28 +158,1187 @@ impl InterpreterStorage for VmDatabase {
...
+                        "current_key overflowed during computation"
+                    )))?;
+        }
+
+        transaction.commit()?;
+
+        // let get_result = self.database.get::<Bytes32>(
+        //     MultiKey::new(&(contract_id, start_key)).as_ref(),
+        //     Column::ContractsState,
+        // );
```
Comment (Voxelot) [inline fuel-core/src/database/vm_database.rs:279]: unused code

```suggestion
```

**Review by xgreenx — COMMENTED** [review]
It was hard to read some parts, so I simplified them https://github.com/FuelLabs/fuel-core/pull/830=)

File: fuel-core/src/database/vm_database.rs [hunk 3]
```diff
@@ -142,28 +158,1179 @@ impl InterpreterStorage for VmDatabase {
...
+        let mut found_unset = false;
+
+        let mut current_key = U256::from_big_endian(start_key.as_ref());
+        let mut key_bytes = [0u8; 32];
+
+        let transaction = self.database.transaction();
+        let transaction_db = transaction.deref();
+
+        for value in values {
+            current_key.to_big_endian(&mut key_bytes);
```
Comment (freesig) [inline fuel-core/src/database/vm_database.rs:258]: Is this actually correct? Won't rocks db be storing the `Bytes32` keys ordered by the actual bytes in the array not as a number?

Comment (Voxelot) [reply fuel-core/src/database/vm_database.rs:229]: The compiler will decide which endianess the keys are in when incrementing slot keys, and they confirmed that they plan to use big endian


cc @mohammadfawaz

**Review by freesig — APPROVED** [review]
All looks good

**Review by Voxelot — APPROVED** [review]

**Pull request merged**